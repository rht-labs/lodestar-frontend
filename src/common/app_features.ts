// All roles defined by the backend to be used in the frontend should be enumerated in this object.

export const APP_FEATURES = {
  admin: 'admin',
  writer: 'writer',
  reader: 'reader',
  notifications: 'notifications',
  engagementCardIcons: 'engagementCardIcons',
  resetUser: 'resetUser',
};

export type AppFeature = keyof typeof APP_FEATURES;
