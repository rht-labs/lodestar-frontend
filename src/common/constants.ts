export const APPLICATION_CONTENT_CONTAINER_ID = 'main-application-content';
