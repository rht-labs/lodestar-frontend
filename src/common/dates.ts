import { format as formatDate, utcToZonedTime } from 'date-fns-tz';
export const formatUtcDate = (date?: Date): string | undefined => {
  if (!date) {
    return '';
  }
  const dateString = formatDate(utcToZonedTime(date, 'UTC'), 'MMM dd, yyyy', {
    timeZone: 'UTC',
  });
  return dateString;
};
