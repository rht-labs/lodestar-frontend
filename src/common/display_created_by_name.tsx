import React from 'react';

export function DisplayCreatedByName({
  userFromServer,
  lastUpdatedBy,
}: {
  userFromServer?: string;
  lastUpdatedBy?: string;
}) {
  return <>{!!userFromServer ? userFromServer : lastUpdatedBy}</>;
}
