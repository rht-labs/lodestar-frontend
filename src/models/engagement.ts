export interface Engagement {
  name: string;
}
export class Engagement {
  constructor({ name }: Engagement) {
    this.name = name;
  }
  name: string;
}
