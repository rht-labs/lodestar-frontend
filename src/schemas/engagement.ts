import faker from 'faker/locale/en_US';
import { LaunchData } from './launch_data';
import { GitCommit } from './git_commit';
import { CreationDetails } from './creation_details';
import { ClusterStatus } from './cluster_status';
export interface Engagement {
  additional_details?: string;
  archive_date: Date;
  commits: GitCommit[];
  customer_contact_email: string;
  customer_contact_name: string;
  customer_name: string;
  description: string;
  end_date: Date;
  engagement_lead_email: string;
  engagement_lead_name: string;
  engagement_users: Array<any>;
  last_update: string;
  location: string;
  mongo_id?: string;
  ocp_cloud_provider_name: string;
  ocp_cloud_provider_region: string;
  ocp_cluster_size: string;
  ocp_persistent_storage_size: string;
  ocp_sub_domain: string;
  ocp_version: string;
  project_id: number;
  project_name: string;
  start_date: Date;
  technical_lead_email: string;
  technical_lead_name: string;
  launch?: LaunchData;
  creation_details: CreationDetails;
  last_update_by_name: string;
  suggested_subdomain?: string;
  status: ClusterStatus;
}
export abstract class Engagement {
  static fromFake(staticData = false): Engagement {
    return {
      additional_details: staticData
        ? 'Additional information here'
        : faker.lorem.paragraphs(2),
      archive_date: staticData ? new Date(2020, 6, 30) : faker.date.recent(),
      commits: [],
      customer_contact_email: staticData
        ? 'bob@doe.com'
        : faker.internet.email(),
      customer_contact_name: staticData
        ? 'Bob Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      customer_name: staticData ? 'NASA' : faker.company.companyName(),
      description: staticData
        ? "It's rocket science"
        : faker.lorem.paragraphs(5),
      end_date: staticData ? new Date(2020, 6, 1) : faker.date.future(),
      engagement_users: staticData ? [] : [],
      engagement_lead_email: staticData
        ? 'alice@doe.com'
        : faker.internet.email(),
      engagement_lead_name: staticData
        ? 'Alice Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      last_update: staticData ? '2020-01-01' : new Date().toString(),
      location: staticData
        ? 'Nashville, TN'
        : `${faker.address.city()}, ${faker.address.stateAbbr()}`,
      mongo_id: '1',
      ocp_cloud_provider_name: staticData ? 'AWS' : 'AWS',
      ocp_cloud_provider_region: staticData
        ? 'N. Virginia'
        : faker.lorem.word(),
      ocp_cluster_size: staticData ? 'Large' : 'Large',
      ocp_persistent_storage_size: staticData
        ? '100G'
        : `${faker.random.number()}G`,
      ocp_sub_domain: staticData ? 'ordoabchao' : faker.lorem.word(),
      ocp_version: staticData ? '4.4' : faker.random.number().toString(),
      project_id: staticData ? 1 : faker.random.number(),
      project_name: staticData ? 'Boots on the Moon' : faker.company.bsNoun(),
      start_date: staticData ? new Date(2020, 5, 1) : faker.date.recent(),
      technical_lead_email: staticData ? 'eve@doe.com' : faker.internet.email(),
      technical_lead_name: staticData
        ? 'Eve Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      creation_details: staticData
        ? {
            created_by_email: 'dwasinger@redhat.com',
            created_by_user: 'dwasinge',
            created_on: new Date(2020, 1, 1),
          }
        : {
            created_by_email: faker.internet.email(),
            created_by_user: `${faker.name.firstName()} ${faker.name.lastName()}`,
            created_on: faker.date.recent(),
          },
      last_update_by_name: staticData
        ? 'James Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      launch: staticData
        ? null
        : faker.random.boolean()
        ? {
            launched_by: faker.name.firstName(),
            launched_date_time: faker.date.recent(),
          }
        : null,
      status: ClusterStatus.fromFake(staticData),
    };
  }
}

export enum EngagementStatus {
  active = 'active',
  past = 'past',
  upcoming = 'upcoming',
}

export const getEngagementStatus = (
  engagement: Engagement
): EngagementStatus => {
  if (!engagement) {
    return null;
  }
  const Today = new Date();
  const { launch, end_date } = engagement;
  const hasLaunched = !!launch?.launched_date_time;
  if (hasLaunched && end_date >= Today) {
    return EngagementStatus.active;
  } else if (hasLaunched && end_date < Today) {
    return EngagementStatus.past;
  } else {
    return EngagementStatus.upcoming;
  }
};
