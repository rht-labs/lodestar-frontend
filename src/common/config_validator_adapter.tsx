import { ValidatorFactory } from '../schemas/validators';
import { EngagementFormConfig } from '../schemas/engagement_config';

export const getValidatorsFromFormOptions = (
  formOptions: EngagementFormConfig = {}
) =>
  Object.keys(formOptions || {}).reduce(
    (previousSchemaGroups, currentGroupKey) => {
      return {
        ...previousSchemaGroups,
        ...Object.keys(formOptions[currentGroupKey] ?? {}).reduce(
          (previousGroupFields, currentFieldKey) => ({
            ...previousGroupFields,
            [currentFieldKey]: (
              formOptions?.[currentGroupKey]?.[currentFieldKey]?.validators ||
              []
            ).map(ValidatorFactory),
          }),
          {}
        ),
      };
    },
    {}
  );
