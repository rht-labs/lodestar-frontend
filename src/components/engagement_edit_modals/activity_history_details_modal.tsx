import React from 'react';
import {
  Modal,
  ModalVariant,
  Button,
  Grid,
  GridItem,
} from '@patternfly/react-core';
import { useModalVisibility } from '../../context/edit_modal_visibility_context/edit_modal_visibility_hook';
import { EditModalTemplate } from '../../layout/edit_modal_template';
import { GitCommit } from '../../schemas/git_commit';
import { ActivityHistoryLineItem } from '../activity_history_line_item/activity_history_line_item';
import { Engagement } from '../../schemas/engagement';
export interface ActivityHistoryDetailsModalProps {
  isOpen: boolean;
  engagement: Engagement;
}
export function ActivityHistoryDetailsModal(
  props: ActivityHistoryDetailsModalProps
) {
  const { requestClose } = useModalVisibility();
  const anyActivities = (props.engagement?.commits?.length > 0);
  return (
    <Modal
      variant={ModalVariant.small}
      isOpen={props.isOpen}
      onClose={requestClose}
      title="Activity History"
    >
      <EditModalTemplate
        actions={
          <div>
            <Button onClick={requestClose}>Close</Button>
          </div>
        }
      >
        {
          anyActivities
          ? <DetailedActivityHistoryList commits={props.engagement?.commits} />
          : <p style={{fontStyle: 'italic'}}> No activity to display</p>
        }
      </EditModalTemplate>
    </Modal>
  );
}

function DetailedActivityHistoryList({ commits }: { commits: GitCommit[] }) {
  return (
    <Grid hasGutter>
      {commits?.map(commit => (
        <GridItem span={12}>
          <ActivityHistoryLineItem commit={commit} />
        </GridItem>
      ))}
    </Grid>
  );
}
