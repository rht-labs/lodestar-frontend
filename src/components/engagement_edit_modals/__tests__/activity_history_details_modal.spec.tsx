import React from 'react';
import { render } from '@testing-library/react';
import { TestStateWrapper } from '../../../common/test_state_wrapper';
import { ActivityHistoryDetailsModal } from '../activity_history_details_modal';
import { Engagement } from '../../../schemas/engagement';

describe('Activity History Details Modal', () => {
  test('matches snapshot', () => {
    expect(
      render(
        <TestStateWrapper>
          <ActivityHistoryDetailsModal
            engagement={Engagement.fromFake(true)}
            isOpen={true}
          />
        </TestStateWrapper>
      )
    );
  });
});
