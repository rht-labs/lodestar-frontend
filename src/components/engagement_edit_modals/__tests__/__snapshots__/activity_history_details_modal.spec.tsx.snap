// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity History Details Modal matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="pf-c-backdrop__open"
  >
    <div
      aria-hidden="true"
    />
    <div>
      <div
        class="pf-c-backdrop"
      >
        <div
          class="pf-l-bullseye"
        >
          <div
            aria-describedby="pf-modal-part-2"
            aria-labelledby="pf-modal-part-1"
            aria-modal="true"
            class="pf-c-modal-box pf-m-sm"
            data-ouia-component-id="0"
            data-ouia-component-type="PF4/ModalContent"
            data-ouia-safe="true"
            id="pf-modal-part-0"
            role="dialog"
          >
            <button
              aria-disabled="false"
              aria-label="Close"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                  transform=""
                />
              </svg>
            </button>
            <header
              class="pf-c-modal-box__header"
            >
              <h1
                class="pf-c-modal-box__title pf-c-modal-box__title"
                id="pf-modal-part-1"
              >
                Activity History
              </h1>
            </header>
            <div
              class="pf-c-modal-box__body"
              id="pf-modal-part-2"
            >
              <div
                data-testid="activity_history_modal"
              >
                <section
                  class="pf-c-page__main-section"
                >
                  <dl
                    aria-label=""
                    class="pf-c-accordion"
                  >
                    <div>
                      <button
                        aria-expanded="false"
                        class="pf-c-accordion__toggle"
                        id="1"
                      >
                        <span
                          class="pf-c-accordion__toggle-text"
                        >
                          <div
                            class="pf-l-flex"
                          >
                            <div
                              class=""
                            >
                              <span>
                                1 Feb  2020, 00:00 UTC-05
                              </span>
                            </div>
                            <div
                              class=""
                            >
                              <span>
                                A commit message
                              </span>
                            </div>
                            <div
                              class=""
                            >
                              random-user@example.com
                            </div>
                          </div>
                        </span>
                        <span
                          class="pf-c-accordion__toggle-icon"
                        >
                          <svg
                            aria-hidden="true"
                            fill="currentColor"
                            height="1em"
                            role="img"
                            style="vertical-align: -0.125em;"
                            viewBox="0 0 256 512"
                            width="1em"
                          >
                            <path
                              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                              transform=""
                            />
                          </svg>
                        </span>
                      </button>
                    </div>
                    <dd
                      aria-label=""
                      class="pf-c-accordion__expanded-content"
                      hidden=""
                      id=""
                    >
                      <div
                        class="pf-c-accordion__expanded-content-body"
                      >
                        <div>
                          <div>
                            <span
                              style="white-space: pre-wrap;"
                            >
                              A commit message
                            </span>
                          </div>
                          <div>
                            <a
                              href="https://gitlab.example.com/my-org/engagements/company/project/iac/-/commit/e169e7c04a9c180a7a75af124f0cbfc657fc47ae"
                            >
                              See in GitLab
                            </a>
                          </div>
                        </div>
                      </div>
                    </dd>
                  </dl>
                </section>
                <section
                  class="pf-c-page__main-section"
                >
                  <div
                    style="margin-top: 1rem;"
                  >
                    <div
                      class="pf-l-flex pf-m-justify-content-flex-end"
                    >
                      <div
                        class=""
                      >
                        <div>
                          <button
                            aria-disabled="false"
                            class="pf-c-button pf-m-primary"
                            data-ouia-component-id="2"
                            data-ouia-component-type="PF4/Button"
                            data-ouia-safe="true"
                            type="button"
                          >
                            Close
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </section>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
